### ISSUE TITLE
Enable OSS community to write their own device drivers

### ISSUE THREAD

**Issue by DanielMSchmidt** [wix/Detox#1091]
### Problem

Every single platform we want to support (e.g. iOS Siumlator, iOS Device, AWS Device Farm devices, ...) currently needs to be enabled in this repository. We might run into situations where we have a very good PR for one platform, but we can not support it or we don't want to spend time on supporting it. 

### Proposed Solution

We need to bring in some sort of modularization that will allow us to execute code plugins provide depending on configuration. To solve this specific problem, the idea would be that plugins may provide their own device type, e.g. `iOSDevice` and their own parameters of configuration that might be needed. As a user you would install another npm package e.g. `detox-plugin-real-ios-device`, do the configuration based on the documentation of the plugin and be done.

### Proposed Implementation

We could either use a tool like `gluegun` or simply use a [similar loading mechanism](https://github.com/infinitered/gluegun/blob/7b7fbd11f98100f78fad1844a660edc9efa1b171/src/loaders/plugin-loader.ts#L18). They expect certain folders as entry points within a directory. This way would automatically detect these plugins and would let them seamlessly integrate.

The other option would be to go a similar route like `jest` does and have extensibility baked into the API. Examples for this would be [`expect.extend`](https://jestjs.io/docs/en/expect#expectextendmatchers) or [reporters](https://jestjs.io/docs/en/configuration.html#reporters-array-modulename-modulename-options).

### Challenges

- We should not introduce unnecessary breaking changes in the process
- We need to find an extension pattern that suits our needs in general as there might be other parts like matchers or expectations which we also would like to make extensible later on

**Comment by LeoNatan** [comment]
This is a good idea.

We need to understand where the limits of such a system are. I don't think it is possible to consider entirely new systems (such real iOS device) as something that can be achieved with such an extensibility support. Unless we rewrite Detox from the core, which I don't think we will.

The JS is a small part of Detox, and while important as orchestrator of the lifecycle, it is still a small part. Most hard work for adding such a support usually happens in the native side.

I do think that such extensability systems are useful to enhance the existing systems so as to achieve additional functionality (for example, appium support). Or support systems that are implemented entirely elsewhere, such as using Appium as a "driver" but actually being thin wrappers around external APIs.

**Comment by LeoNatan** [comment]
I don't think we should be loading third party code. We should allow public APIs for registering points of integration ("register this class to handle that configuration type") or ability to give "enhancer" objects which somehow influence the way existing logic is performed ("take this object, which implements that protocol/contract, so that it can influence how we perform that action").

**Comment by LeoNatan** [comment]
I think we should also think how many actual cases exist for this enhancement. While important, I fear there may not be as much necessity. One big example I can think of is https://github.com/wix/Detox/pull/1049 which could be implemented like that if we had support. I don't think 'real iOS Device' fits.

Maybe we can split this into several different projects and implement on demand (for example, implement custom configuration types first because we need it now, but implement other integration later, as needed).

Of course, we need to have an idea first what we want to implement before we implement :)

**Comment by EdwardDrapkin** [comment]
@LeoNatan I think the biggest argument for doing this (or something like it) is what you said in #1049: maintainability.  If there are plugins that can provide drivers, than detox-appium, detox-aws-farm, etc. can be third party packages that you may or may not have to be involved with.  

With that in mind, I think it might be worth considering an approach like this:

1. Formalize a Javascript interface for device drivers and interactions.  This is already mostly done in the code because of the necessity that android/ios implementations look the same.

2. Create a loading mechanism that allows a single "device provider" to be specified in the configuration.  The current field in the config (`android.attached`, `ios.simulator`) could be repurposed without breaking backwards compatibility.

3. Migrate the current implementations to plugin implementations.  The plugins would ship the native code (EarlGrey) along side the javascript "plugins" that detox would interface with.  This opens the door for appium, aws, etc. as well as potential other native drivers that might want to use something other than EarlGrey for whatever reason (future proofing, basically).  

 I've alluded to it, but I want to be specific too: I think this is the most future proof and easy way for you guys.  It would allow the community (which is growing pretty fast!) to take a greater load off of your shoulders because it will make supporting things like appium zero effort for you, and it will make detox itself more accessible for people to make changes and improvements to (and the same goes for individual device drivers).  As far as future proofing, if any of the current assumptions (EarlGray, adb, simctl, etc.) it makes it much easier to move on without disrupting anyone.  

I don't think much actual code would have to be written to get this done, mostly documentation and reorganization.  And I'd be more than willing to help because this path seems like a much clearer and more fruitful path to my ultimate goal of running on AWS than trying to get the tests running on an iOS device alone in a cave.

**Comment by DanielMSchmidt** [comment]
@LeoNatan Thank you for your input, it was incredibly valuable 👍 I agree with your points and I would see this ticket as the very first project, allowing the extension of configuration. Once we have this we can move on with other areas that might be interesting to explore, but first things first.


@EdwardDrapkin Thank you for documenting down your thoughts, I will take a lot of inspiration from them 👍 I think shipping native code might be feasible someday and I think you would agree if I say we should go there in small & continuous steps, making sure we ship what the community actually needs.


Just to share the current state: My first goal is to refactor how we do configurations currently, so it's an easier to extend interface later on. I plan to ship this as an individual refactoring PR preparing this work.

**Comment by LeoNatan** [comment]
@EdwardDrapkin Running tests on an iOS device alone in the cave is a major prerequisite to running them on real devices on AWS. Solving the local case usually helps solving the generalized case. 

In general, I agree. Being generic is a good idea. In practice, creating a native infrastructure for loading plugins is not something I want us to tackle at this point. It is such a complex topic, with so many pitfalls. I want us to concentrate the few resources we have to support the stated use cases in the readme: iOS and Android support. Those "small" use-cases give us enough trouble as it is, without complicating the system with more and more abstractions. I don't see a point in focusing on these abstractions when the stated purpose of Detox is not perfected or close to it.

I would like to look at things honestly here. The use cases we've seen from the community that require this advanced architecture are few. I don't remember one case where an entire new use-case was needed. People want to use what Detox is good at, which is synchronizing native and react native actions between tests. For example, in the use case of running on a real iOS device; would you want to implement everything Detox does on your own, in order to support such devices? Or would you prefer to have the existing Detox for iOS implementation work there too? If the latter, then the proposed abstraction doesn't work, because simulator and real device are then tightly coupled. As are Android emulators and real devices. If you can implement the former, do you really need Detox?

As I said in my initial thoughts, creating a wholly new use-case, such as Appium or, theoretically, [Windows](https://github.com/Microsoft/react-native-windows), those cases make much more sense to look at currently, and implementing them would be an order of magnitude more simpler ("here is an interface, implement it, give us—somehow—an instance that implements it and we'll use that instead of ours").

At the end of the day, I think focus is key. I want us to be focused on the purpose of this project. I am not sure that creating an Appium-style, general purpose test runner or test management solution. Detox is unique in its support for the stated platforms and first-class native and react native support on Android and iOS. I'd prefer we keep most of our focus there until a certain threshold of quality is reached.

**Comment by LeoNatan** [comment]
Regarding helping with Detox; in the other issue I have outlined my progress and what issues I foresee before we can run on iOS devices. Contributions are welcome. I don't see how you'd run on an AWS farm without solving those, be it in Detox or a plugin.

**Comment by DanielMSchmidt** [comment]
@LeoNatan Let me try to summarize your concerns to make sure I really understand them. 

You would not want us to go into a direction of becoming a very broad tool instead of keeping the focus on the things we do well. You would like to keep the new API surface exposed as explicit, minimal and precise as possible to minimize the maintainability and development effort for this feature as it's quite specific to a small audience. Also, you pointed out that having real device support as a plugin will most probably not be feasible.

Would you say I got your concerns? I'm trying to make sure I truly understand you so that I can be sure to evaluate my solutions against these constraints and we get the best outcome possible given the time constraint.

**Comment by LeoNatan** [comment]
Yes. At least for the time being.

I think we should provide a solution for the Appium request; this can work to test the water with some extensibility.

We should also eventually work on supporting real devices, but not as effort for extensibility but because we want this feature as part of Detox. We don't have resources for this right now (I am busy with Detox Instruments for the time being).

**Comment by DanielMSchmidt** [comment]
Okay, great 👍 I agree and I'll explore the problem space a bit more and ping y'all back with an API draft somewhen next year (this is so easy to say right now :D)

**Comment by LeoNatan** [comment]
❤

Happy holidays!

**Comment by DanielMSchmidt** [comment]
Hey folks,

I have thought about this a little and here is the API I would like to suggest:


```diff
```

**Comment by OrEvorn** [comment]
Hi all,
any updates on this one? :)

**Comment by LeoNatan** [comment]
This is currently out of scope for the project. We are spread thin as it is. Closing for now.
